### PR TITLE
Make the SMTP Server follow the spec

### PR DESCRIPTION
**Pull request by cfpwastaken**
Currently, some things in the SMTP Server do not exactly match the spec.

### CODE REVIEW

**Review by j0code — COMMENTED** [review]
yeah, I'm doing something better atm, so close ig

Edit: don't close, changes requested below

**Comment by cfpwastaken** [comment]
what do you mean

File: src/smtp/SMTPServer.ts [hunk 1]
```diff
@@ -9,6 +9,28 @@ import crypto from "node:crypto";
...
+		[550, "Requested action not taken: mailbox unavailable"], // when we cant find the user
+		[451, "Requested action aborted: error in processing"],
+		[551, "User not local; please try <forward-path>"],
+		[452, "Requested action not taken: insufficient system storage"],
+		[552, "Requested mail action aborted: exceeded storage allocation"], // when the user has exceeded their quota
+		[553, "Requested action not taken: mailbox name not allowed"], // when the user has an invalid name
+		[354, "Start mail input; end with <CRLF>.<CRLF>"],
+		[554, "No valid recipients"],
+		[250, "OK"],
+	]);
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:33]: remove

File: src/smtp/SMTPServer.ts [hunk 2]
```diff
@@ -37,7 +59,7 @@ export default class SMTPServer {
 					receivingData = false;
 					info.content = info.content.substring(0, info.content.length - 3).replaceAll("\r\n", "\n")
 					await smtpserver.handleNewMail(info)
-					sock.write("250 OK\r\n")
+					sock.write(`250 ${this.messages.get(250)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:62]: undo

File: src/smtp/SMTPServer.ts [hunk 3]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
+				info = {
+					from: "",
+					to: [],
+					content: ""
+				}
 				const email = msg.split(":")[1].split(">")[0].replace("<", "")
 				console.log("[SMTP] MAIL FROM: " + email)
 				info.from = email
-				sock.write("250 OK\r\n")
+				sock.write(`250 ${this.messages.get(250)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:83]: undo

File: src/smtp/SMTPServer.ts [hunk 4]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
+				}
+				const user = await User.findOne({ where: { username } })
+				if(!user) {
+					sock.write(`550 ${this.messages.get(550)}\r\n`)
+					return
+				}
 				info.to.push(email)
 				console.log("[SMTP] RCPT TO: " + email)
-				sock.write("250 OK\r\n")
+				sock.write(`250 ${this.messages.get(250)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:105]: undo

File: src/smtp/SMTPServer.ts [hunk 5]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
+				}
 				const email = msg.split(":")[1].split(">")[0].replace("<", "")
 				console.log("[SMTP] MAIL FROM: " + email)
 				info.from = email
-				sock.write("250 OK\r\n")
+				sock.write(`250 ${this.messages.get(250)}\r\n`)
 			} else if(msg.startsWith("RCPT TO:")) {
+				if(info.from == "") {
+					// The spec says we should return 503 if the client has not sent MAIL FROM yet
+					sock.write(`503 ${this.messages.get(503)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:87]: undo

File: src/smtp/SMTPServer.ts [hunk 6]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
+					return
+				}
 				const email = msg.split(":")[1].split(">")[0].replace("<", "")
+				const username = email.split("@")[0]
+				const domain = email.split("@")[1]
+				if(domain != getConfig("host")) {
+					// The spec says we MAY forward the message ourselves, but simply returning 550 is fine, and the client should handle it
+					sock.write(`550 ${this.messages.get(550)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:95]: undo

File: src/smtp/SMTPServer.ts [hunk 7]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
+					return
+				}
+				const user = await User.findOne({ where: { username } })
+				if(!user) {
+					sock.write(`550 ${this.messages.get(550)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:100]: undo

File: src/smtp/SMTPServer.ts [hunk 8]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
+				}
 				info.to.push(email)
 				console.log("[SMTP] RCPT TO: " + email)
-				sock.write("250 OK\r\n")
+				sock.write(`250 ${this.messages.get(250)}\r\n`)
 			} else if(msg.startsWith("DATA")) {
+				// The spec says we should return either 503 or 554 if the client has not sent MAIL FROM or RCPT TO yet
+				// We will send 554 because it is more specific
+				if(info.from == "" || info.to.length == 0) {
+					sock.write(`554 ${this.messages.get(554)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:110]: undo

File: src/smtp/SMTPServer.ts [hunk 9]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
+					return
+				}
 				receivingData = true;
 				console.log("[SMTP] Now receiving data -----------------------------------")
-				sock.write("354 Start mail input; end with <CRLF>.<CRLF>\r\n")
+				sock.write(`354 ${this.messages.get(354)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:115]: undo

File: src/smtp/SMTPServer.ts [hunk 10]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
 			} else if(msg.startsWith("QUIT")) {
-				sock.write("221 Bye\r\n")
+				sock.write(`221 ${this.messages.get(221)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:117]: undo

File: src/smtp/SMTPServer.ts [hunk 11]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
 				sock.end()
+			} else if(msg.startsWith("VRFY")) {
+				// This command is used to verify if a user exists, but that can be a security risk + it is also done with RCPT TO anyway
+				sock.write(`502 ${this.messages.get(502)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:121]:
```suggestion
				sock.write("502 Command not implemented\r\n")
```

File: src/smtp/SMTPServer.ts [hunk 12]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
+			} else if(msg.startsWith("EXPN")) {
+				sock.write(`502 ${this.messages.get(502)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:123]:
```suggestion
				sock.write("502 Command not implemented\r\n")
```

File: src/smtp/SMTPServer.ts [hunk 13]
```diff
@@ -49,24 +71,58 @@ export default class SMTPServer {
...
 			} else {
-				sock.write("500 Command not implemented\r\n")
+				sock.write(`500 ${this.messages.get(500)}\r\n`)
```
Comment (j0code) [inline src/smtp/SMTPServer.ts:125]: undo

**Review by j0code — CHANGES_REQUESTED** [review]
everything I did not comment looks good and will be approved

**Review by j0code — APPROVED** [review]
approved, potential conflicts are your responsibility to resolve

**Comment by cfpwastaken** [comment]
will do, i will merge changes into your branch.

**Pull request merged**